pkg/iter/sample_iterator.go

```diff
@@ -567,14 +568,14 @@ func NewMultiSeriesIterator(series []logproto.Series) SampleIterator {
 // NewSeriesIterator iterates over sample in a series.
 func NewSeriesIterator(series logproto.Series) SampleIterator {
 	return &seriesIterator{
-		i:      -1,
+		i:      atomic.NewInt32(-1),
 		series: series,
 	}
 }
 
 func (i *seriesIterator) Next() bool {
-	i.i++
-	return i.i < len(i.series.Samples)
+	i.i.Inc()
+	return int(i.i.Load()) < len(i.series.Samples)
```

Comment:
I'm not sure if this is correct :thinking:  I don't know exactly how this `seriesIterator` is used, but there can be a data race here, if something calls `Next()` quickly multiple times in parallel, the value of `i` can change between `i.i.Inc()` and `i.i.Load()`, so only the second value will be loaded. Tough now, because of atomics, the go data race detector won't be able to tell there is a problem.

The correct way to load this value would be to save and use the result from `i.Inc()`, since it returns the new value it incremented the variable to: https://pkg.go.dev/go.uber.org/atomic#Int32.Inc

Or, if you use the built-in atomic, `Add()` also returns the new value: https://pkg.go.dev/sync/atomic#Int32.Add

Reply:
Thanks fot catching, ofc calls write and read should not be separate operations. 
Addressed in https://github.com/grafana/loki/pull/12225